android/src/main/java/com/pagopa/securestorage/SecureStorage.kt

```diff
@@ -0,0 +1,400 @@
+package com.pagopa.securestorage
+
+import android.content.Context
+import android.content.pm.PackageManager
+import android.os.Build
+import android.os.storage.StorageManager
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyInfo
+import android.security.keystore.KeyProperties
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_STRONGBOX
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_TRUSTED_ENVIRONMENT
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_UNKNOWN_SECURE
+import android.util.AtomicFile
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
+import java.io.UnsupportedEncodingException
+import java.net.URLDecoder
+import java.net.URLEncoder
+import java.security.KeyStore
+import javax.crypto.Cipher
+import javax.crypto.KeyGenerator
+import javax.crypto.SecretKey
+import javax.crypto.SecretKeyFactory
+import javax.crypto.spec.GCMParameterSpec
+
+
+/**
+ * A secure storage class for writing on encrypted files.
+ * This class has a private constructor, thus instances must be created by using its [Builder].
+ * @property context the application context.
+ * @property storageDirectory the directory where the files will be stored.
+ * @property useEncryption if manual encryption has to be used.
+ */
+class SecureStorage private constructor(
+  context: Context, storageDirectory: File, useEncryption: Boolean
+) {
+
+  private val context: Context
+  private val storageDirectory: File
+  private val useEncryption: Boolean
+
+  init {
+    this.context = context
+    this.storageDirectory = storageDirectory
+    this.useEncryption = useEncryption
+  }
+
+  /**
+   * Removes the file associated with the given [key].
+   * @param key the identifier of the file.
+   * @throws SecureStorageException if an error occurs while deleting the file.
+   */
+  fun remove(key: String) {
+    try {
```

Comment:
Just because we like fp-ts, have a look at `runCatching` 😉. It's not more synthetic than try/catch, but It deserves to be known. What do you think?

Reply:
In this particular case I would say they are equivalent. Using `runCatching` implies also editing the input parameter of `getExceptionMessageOrEmpty` from `Exception` to `Throwable`. I'd keep it this way if that makes to you. Thanks for the suggestion though, another interesting scope function.

Reply:
Yup, sure!